docs/proposals/alternate-api-representations.md

```diff
@@ -0,0 +1,422 @@
+    to ask for v2, then v1)
+* We wish to preserve the greatest possible query parameter space for sub resources
+  and special cases, which encourages us to avoid polluting the API with query
+  parameters that can be otherwise represented.
+* Because we expect not all extensions will implement protobuf, an efficient client
+  must continue to be able to "fall-back" to JSON, such as for third party
+  resources.
+* We do not wish to create fake content-types like `application/json+kubernetes+v1+api.k8s.io`
+  because the list of combinations is unbounded and our ability to encode specific values
+  (like slashes) into the value is limited.
+
+### Client negotiation of response representation
+
+When a client wishes to request an alternate representation of an object, it should form
+a valid `Accept` header containing one or more accepted representations, where each
+representation is represented by a media-type and [media-type parameters](https://tools.ietf.org/html/rfc6838#section-4.3).
+The server should omit representations that are unrecognized or in error - if no representations
+are left after omission the server should return a `406 Not Acceptable` HTTP response.
+
+The supported parameters are:
+
+| Name | Value | Default | Description |
+| ---- | ----- | ------- | ----------- |
+| g | The group name of the desired response | Current group | The group the response is expected in. |
+| v | The version of the desired response | Current version | The version the response is expected in. Note that this is separate from Group because `/` is not a valid character in Accept headers. |
+| as | Kind name | None | If specified, transform the resource into the following kind (including the group and version parameters). |
+| sv | The server group (`api.k8s.io`) version that should be applied to generic resources returned by this endpoint | Matching server version for the current group and version | If specified, the server should transform generic responses into this version of the server API group. |
+| export | `1` | None | If specified, transform the resource prior to returning to omit defaulted fields. Additional arguments allowed in the query parameter. For legacy reasons, `?export=1` will continue to be supported on the request |
+| pretty | `0`/`1` | `1` | If specified, apply formatting to the returned response that makes the serialization readable (for JSON, use indentation) |
+
+For both export and the more complicated server side `kubectl get` cases, it's likely that
+more parameters are required and should be specified as query parameters. However, the core
+behavior is best represented as a variation on content-type. Supporting both is not limiting
+in the short term as long as we can validate correctly.
+
+
+### Example: Partial metadata retrieval
+
+The client may request to the server to return the list of namespaces as a
+`PartialObjectMetadata` kind, which is an object containing only `ObjectMeta` and
```

Comment:
Why not just call it `ObjectMetadata`?
That way any client knows that if it requires a field X from namespaces, it can go a GET /api/v1/namespaces with Accept as=X.
How does it know that it needs to set `as=PartialObjectMetadata` for getting `ObjectMetadata`.